### ISSUE TITLE
[LifetimeSafety] False-positive std::shared_ptr constructed from std::unique_ptr

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#193747]
```cpp
#include <memory>

std::unique_ptr<int> create_up();
std::shared_ptr<int> create_sp() { return create_up(); }
```

Fails with 

```
<source>:4:43: warning: address of stack memory is returned later [-Wlifetime-safety-return-stack-addr]
    4 | std::shared_ptr<int> create_sp() { return create_up(); }
      |                                           ^~~~~~~~~~~
<source>:4:43: note: returned here
    4 | std::shared_ptr<int> create_sp() { return create_up(); }
      |                                           ^~~~~~~~~~~
1 warning generated.
Compiler returned: 0
```

https://godbolt.org/z/xor7fofrE

**Comment by usx95** [comment]
Simpler reproducer without shared_ptr. This is a case of inference where the storage is moved (so this is low confidence inference)

```cpp
#include <memory>

std::unique_ptr<int> create_up();

struct S {
    int* field;
    S(std::unique_ptr<int>&& up) : field(up.get()) { up.release(); }
};

S foo() { return S(create_up()); }
```

```
<source>:7:7: warning: parameter in intra-TU function should be marked [[clang::lifetimebound]] [-Wlifetime-safety-intra-tu-suggestions]
    7 |     S(std::unique_ptr<int>&& up) : field(up.get()) { up.release(); }
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                 [[clang::lifetimebound]]
<source>:6:10: note: escapes to this field
    6 |     int* field;
      |     ~~~~~^~~~~
<source>:10:20: warning: returning address of local temporary object [-Wreturn-stack-address]
   10 | S foo() { return S(create_up()); }
      |                    ^~~~~~~~~~~
<source>:10:20: warning: address of stack memory is returned later [-Wlifetime-safety-return-stack-addr]
   10 | S foo() { return S(create_up()); }
      |                    ^~~~~~~~~~~
<source>:10:18: note: returned here
   10 | S foo() { return S(create_up()); }
      |                  ^~~~~~~~~~~~~~
3 warnings generated.
Compiler returned: 0
```
https://godbolt.org/z/d6nrPrT1P

The `up.release();` should give us a strict warning (alias before moved).
I think the solution would be to infer and suggest annotation only in high confidence cases of escape detection.
